test/functional/feature_pop_stratum_compat.py

```diff
@@ -0,0 +1,122 @@
+"""
+
+from test_framework.pop import endorse_block, create_endorsed_chain, mine_until_pop_active
+from test_framework.authproxy import JSONRPCException
+from test_framework.pop_const import POP_PAYOUT_DELAY
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import (
+    connect_nodes,
+    disconnect_nodes, assert_equal, assert_raises_rpc_error,
+)
+import json
+
+def assert_no_field(source, key):
+    assert isinstance(source, dict)
+    if key in source:
+        raise Exception("Key {} expected to be missing in {}, but it is present: {}".format(key, source, source[key]))
+
+def assert_field_exists(source, key, type):
+    assert isinstance(source, dict)
+    if key not in source:
+        raise Exception("Expected key {} to exist in {}".format(key, source))
+
+    try:
+        if not type(source[key]):
+            raise Exception("Key {} type verification failed. Value={}".format(key, source[key]))
+    except Exception as e:
+        print(e)
+        raise
+
+'''
+If this test fails, then compatibility with Stratum (https://github.com/VeriBlock/vbk-ri-stratum-pool) is broken.
+'''
+class PopStratumCompat(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 1
+        self.extra_args = [["-txindex"]]
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+        self.skip_if_no_pypoptools()
+
+    def setup_network(self):
+        self.setup_nodes()
+
+    def getblocktemplate(self):
+        arg = {
+            "capabilities": ["coinbasetxn", "workid", "coinbase/append"],
+            "rules": ["segwit"]
+        }
+        return self.nodes[0].getblocktemplate(arg)
+
+    def run_test(self):
+        """Main test logic"""
+        from pypoptools.pypopminer import MockMiner
+        self.apm = MockMiner()
+
+        node = self.nodes[0]
+        addr = node.getnewaddress()
+
+        # stop 1 block behind activation
+        self.log.info("Mining blocks until activation -5 blocks")
+        mine_until_pop_active(node, addr, delta=-5)
+
+        # check that getblocktemplate does NOT have pop-related fields before POP activation
+
+        self.log.info("Check that getblocktemplate does not have POP fields")
+        resp = self.getblocktemplate()
+        assert_no_field(resp, 'pop_data')
+        assert_no_field(resp, 'pop_data_root')
+        assert_no_field(resp, 'pop_first_previous_keystone')
+        assert_no_field(resp, 'pop_second_previous_keystone')
+        assert_no_field(resp, 'pop_rewards')
+
+        self.log.info("Mining blocks until activation +5 blocks")
+        mine_until_pop_active(node, addr, delta=+5)
+
+        self.log.info("Mine chain of {} consecutive endorsed blocks".format(POP_PAYOUT_DELAY))
+        create_endorsed_chain(node, self.apm, POP_PAYOUT_DELAY, addr)
+        endorse_block(self.nodes[0], self.apm, node.getblockcount() - 5, addr)
```

Comment:
I recommend to log on each step what heights are we at. Here: `mine_until_pop_active(node, addr, delta=-5)`, here: `mine_until_pop_active(node, addr, delta=+5)` and here: `endorse_block(self.nodes[0], self.apm, node.getblockcount() - 5, addr)`.
It may help to debug some rare combinations of POP_PAYOUT_DELAY and POP_ACTIVATION_HEIGHT.